[Agent Builder] MCP tool and bulk import documentation (#4668)
elastic/docs-content: explore-analyze/ai-features/agent-builder/tools/mcp-tools.md

---
navigation_title: "MCP tools"
applies_to:
  stack: preview 9.3
  serverless: preview
---

# Model Context Protocol (MCP) tools

Agent Builder MCP tools enable calling a remote [MCP server's](https://modelcontextprotocol.io/docs/learn/server-concepts) tools in your agent [chat](../chat.md). When you call an Agent Builder MCP tool, it calls the associated tool on the MCP server and returns its result.

## Prerequisites

To use external MCP tools, you first need to set up an [MCP Kibana Stack Connector](kibana://reference/connectors-kibana/mcp-action-type.md). This interface enables Agent Builder MCP tools to communicate with a remote MCP server.

## Adding MCP tools

> ### Adding a single tool

Reviewer: ```suggestion
### Add a single tool
```

> Once an MCP Kibana Stack Connector has been set up, an MCP tool can be created by clicking "New tool" on the [Tools](/explore-analyze/ai-features/agent-builder/tools.md) landing page.

Reviewer: ```suggestion
Once you've set up an MCP Kibana Stack Connector, click "New tool" on the [Tools](/explore-analyze/ai-features/agent-builder/tools.md) landing page and select the `MCP` tool type.
```

:::{image} ../images/mcp-createnewtool-config-example.png
:screenshot:
:alt: Example configuration for a new MCP tool with the Context7 MCP server.
:width: 800px
:::

MCP tools have the following configuration:

MCP Server
:   The MCP Kibana Stack Connector to interface with.

Tool
:   The specific tool on MCP server to create an Agent Builder MCP tool for.

Once a tool is selected, the `Description` field of the tool automatically populates with the description provided by the MCP server.

> ### Bulk import MCP tools

Reviewer: for this section I'd do something like this:

``````
### Bulk import MCP tools

To import multiple tools at once, go to the [Tools](/explore-analyze/ai-features/agent-builder/tools.md) landing page and select **Bulk import MCP tools** from the "Manage MCP" dropdown.

:::{image} ../images/mcp-bulkimport-location.png
:screenshot:
:alt: The Manage MCP dropdown on the Tools landing page with Bulk import MCP tools highlighted.
:width: 500px
:::

Configure the following fields:

MCP Server
:   The MCP Kibana Stack Connector to interface with.

Tools to import
:   The tools from the MCP server to import.

Namespace
:   A prefix for organizing imported tools. Must start with a letter and contain only lowercase letters, numbers, and hyphens.

:::{image} ../images/mcp-bulkimport-config-example.png
:screenshot:
:alt: Example configuration for bulk importing MCP tools from the Context7 MCP server.
:width: 800px
:::

After clicking **Import tools**, Agent Builder creates an MCP tool for each selection.

Each tool's ID is generated as `namespace.tool-name` (for example, `context7.resolve-library-id`), and descriptions are pulled automatically from the MCP server.
``````

---
navigation_title: "MCP tools"
applies_to:
  stack: preview 9.3
  serverless: preview
---

# Model Context Protocol (MCP) tools

> Agent Builder MCP tools enable calling a remote [MCP server's](https://modelcontextprotocol.io/docs/learn/server-concepts) tools in your agent [chat](../chat.md). When you call an Agent Builder MCP tool, it calls the associated tool on the MCP server and returns its result.

Reviewer 2: ```suggestion
Agent Builder MCP tools enable calling a remote [MCP server's](https://modelcontextprotocol.io/docs/learn/server-concepts) tools in your agent [chat](../chat.md). When your Agent calls an MCP tool, it executes the associated tool on the MCP server and returns its result.
```

Not sure if it's necessary to make this distinction in the docs, so I'll leave the decision up to you!

## Prerequisites

> To use external MCP tools, you first need to set up an [MCP Kibana Stack Connector](kibana://reference/connectors-kibana/mcp-action-type.md). This interface enables Agent Builder MCP tools to communicate with a remote MCP server.

Reviewer 2: I noticed that this is the first mention of the term "Kibana Stack Connector" in the docs and that we don't ever call it that in the Kibana UI (even though internally we call them that). Maybe just `MCP Connector` for this and other instances in this doc?

---
navigation_title: "MCP tools"
applies_to:
  stack: preview 9.3
  serverless:
    elasticsearch: preview
    observability: unavailable

>     security: unavailable

Author: I don't believe they'll be available in Obs and Sec serverless initially either - will double check to be sure

---

# Model Context Protocol (MCP) tools

[Model Context Protocol](https://modelcontextprotocol.io/docs/getting-started/intro) (MCP) tools provide an interface by which Agent Builder can leverage a remote [MCP server's](https://modelcontextprotocol.io/docs/learn/server-concepts) tools during chat sessions. A single Agent Builder MCP tool maps directly to a single MCP server tool. Calling an Agent Builder MCP tool will call the associated tool on the MCP server and return its result.

> ## Prerequisites

Reviewer: Just confirming: do users need to enable the feature in 9.3?

> To create and use MCP tools, an [MCP Kibana Stack Connector](kibana://reference/connectors-kibana/mcp-action-type.md) must first be created. This is the interface by which an Agent Builder MCP tool can communicate with a remote MCP server.

Reviewer: ```suggestion
To use external MCP tools, you first need to set up an [MCP Kibana Stack Connector](kibana://reference/connectors-kibana/mcp-action-type.md). This is the interface by which an Agent Builder MCP tool can communicate with a remote MCP server.
```

## How it works

When an agent calls an MCP tool:

1. It gets the input schema (effectively, what input parameters the tool requires to when called) from the MCP Kibana Stack Connector
2. The MCP server's tool is called with the correct parameters using the configured connector
3. The result is returned to the LLM and interpreted for the user

> There is no post-processing involved between invocation and the result being passed back to the LLM.

Reviewer: minor rejigging for clarity and moved 1:1 mapping fact from intro

## Configuration

MCP tools have the following configuration:

> * **`MCP Server`**: The MCP Kibana Stack Connector to interface with

Reviewer: could use a [definition list](https://elastic.github.io/docs-builder/syntax/definition-lists/) here

ps I added the nested definition to the docs based on your example from the stack connector PR :)

---
navigation_title: "MCP tools"
applies_to:
  stack: preview 9.3
  serverless:
    elasticsearch: preview
    observability: unavailable

>     security: unavailable

Reviewer: ```suggestion
  serverless: preview
```
we can simplify this now that we confirmed availability
